### ISSUE TITLE
Distorted links on state machine diagrams

### ISSUE THREAD

**Issue by devaigergely81** [ELTE-Soft/txtUML#250]
The new/layout-sm-diagrams branch is dedicated to the layouting of generated state machine diagrams according to layout descriptions. The main problem we encountered during the implementation is that links get distorted on the diagrams.

**Comment by devaigergely81** [comment]
@dobreffandras, @GregoBalu: I think I have found the root cause of this problem... For some reason, Papyrus decreases the height of all states by 20 pixels, when our generated diagrams are opened. The notation file contains the correct value, and it is also visible on the diagram for a moment, before the shrinking happens.
In the notation file the states are divided into two parts: When diagrams are edited by hand, the upper part is 20 pixels high and the lower part is the rest. When generating the diagrams and resizing the states using the API, the upper part becomes 40 pixels high. I guess that Papyrus cannot handle this when showing the diagram, and, therefore, it shrinks the upper part (and consequently the state itself) by 20 pixels. If I change the notation file by hand to reset the upper part to 20 pixels and to add 20 pixels to the lower part, the diagram is shown OK. Unfortunately, I was not able to reach the same from the API...
I have finally applied a workaround in the 'deva/sm-diagrams' branch: 20 pixels are added to the height of all states. See https://github.com/ELTE-Soft/txtUML/commit/2abd941e4a733f7edb670a7f05cd26a9367a0dee. The diagrams look much better this way (but still not perfect). Do you have any ideas to solve this in a better way?
I suggest to go forward with the 'deva/sm-diagrams' branch, as it merged master with 'new/layout-sm-diagrams' and contains some additional fixes. Please review it and try to clean up your parts, to be able to create a pull request shortly. Thanks!

**Comment by dobreffandras** [comment]
I've also encountered the problem, that after the diagram generation the states are shrinked by Papyrus. This causes that the lines get distorted. I've tried two different Commands. One of them is a GEF command which resizes a general GraphicalEditPart. Now it uses the 'org.eclipse.papyrus.uml.diagram.statemachine.custom.commands.CustomStateResizeCommand' 
Unfortunatelly using the new, latter command did not solve the problem. If we could find the workflow how states get resized by user interaction we would be closer to the right way.

**Comment by devaigergely81** [comment]
I have created a workaround for this in https://github.com/ELTE-Soft/txtUML/commit/2305dc87d1766b7b60b9a87218613c96ebe3a4b7
I suggest revisiting this issue later, if we can find a better or the related Papyrus bugs are resolved.

**Comment by dobreffandras** [comment]
I was also experimenting and came to the exact same solution that you did. 
On the 23rd I wrote a topic on the eclipse forum. (https://www.eclipse.org/forums/index.php/t/1075040/) 
Until now I did not get any answer. I will write a comment about what we have found out.
Anyway I suggest the same as you, I would close this for now (after cleaning up some code) and reopen it when this bug gets resolved or someone gives a reasonable explanation for the behaviour (and a solution for the problem).
